### PR TITLE
Ansible refactor

### PR DESCRIPTION
**Pull request by kotfic**
Implements an ansible based cluster 'type.'  This is not strictly a type because actually ansible provides the ability to launch instances through many different cloud providers. Currently this only implements launch instances on EC2 but can (and will) be extended to do much more. This branch should not interfere with any previous functionality (e.g. Starcluster based launching). 

@cjh1 PTAL

### CODE REVIEW

File: girder/cumulus/plugin_tests/cluster_test.py [hunk 1]
```diff
@@ -898,31 +897,3 @@ def test_delete_assetstore (self, generate_key):
...
-        r = self.request('/clusters', method='POST',
-                         type='application/json', body=json_body, user=self._user)
-        self.assertStatus(r, 400)
-
-        trad_body['config']['scheduler']['type'] = 'slurm'
-        json_body = json.dumps(trad_body)
-        r = self.request('/clusters', method='POST',
-                         type='application/json', body=json_body, user=self._user)
-        self.assertStatus(r, 201)
-
```
Comment (cjh1) [inline girder/cumulus/plugin_tests/cluster_test.py]: Why is this removed?

File: girder/cumulus/server/cluster.py [hunk 2]
```diff
@@ -160,6 +163,22 @@ def _create_ec2(self, params, body):
 
         return cluster
 
+    def _create_ansible(self, params, body):
+
+        self.requireParams(['name', 'template', 'profile'], body)
+
```
Comment (cjh1) [inline girder/cumulus/server/cluster.py]: 'template' is specfic to starcluster, not sure its needed with ansible?

File: cumulus/ansible/tasks/cluster.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        runner_callbacks=callbacks.PlaybookRunnerCallbacks(stats, verbose=1),
+        stats=stats,
+        extra_vars=extra_vars
+    )
+
+    # Note:  can refer to callback.playbook.extra_vars  to get access
+    # to girder_token after this point
+
+    pb.run()
+
```
Comment (cjh1) [inline cumulus/ansible/tasks/cluster.py]: If running the playbook fails, is an exception raise? If so we probably need a except block to update the cluster state.

File: cumulus/ansible/tasks/playbooks/default.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+          group_desc: "Ec2 security group for cluster {{ cluster_id }}"
+
+      hosts:
+        head:
+          type: "{{ instance_type }}"
+
+        data:
+          type: "{{ instance_type }}"
+          count: "{{ instance_count }}"
+  roles:
```
Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/default.yml]: We will need to be able to specify different instance types for the head and data nodes.

File: cumulus/ansible/tasks/playbooks/filter_plugins/ec2.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        terminate_set |= set(running_list[num_to_keep:])
+
+    return {"start": list(start_set), "terminate": list(terminate_set)}
+
+
+def get_ec2_hosts(instance_table):
+    import operator as op
+    return map(op.itemgetter("id"), instance_table)
+
+
```
Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/filter_plugins/ec2.py]: Where are these filters used? In the default playbook?

File: girder/cumulus/server/cluster.py (hunk 2 above)
Comment (kotfic) [inline girder/cumulus/server/cluster.py]: In this case we use template to pass in extra variables needed by ansible to run the playbook. Theoretically any variable can be passed in,  but practically we just pass in instance type, number of instances etc.

File: girder/cumulus/plugin_tests/cluster_test.py (hunk 1 above)
Comment (kotfic) [inline girder/cumulus/plugin_tests/cluster_test.py]: I think this was probably a rebase mistake,

File: girder/cumulus/server/models/cluster.py [hunk 6]
```diff
@@ -20,16 +20,27 @@
...
+
+def preprocess_cluster(cluster):
+    # Convert model status into enum
+    try:
+        cluster['status'] = ClusterStatus(int(cluster['status']))
+    except (ValueError, AssertionError, TypeError):
+        # Assume 'old style' string status
+        pass
+
+    return cluster
```
Comment (cjh1) [inline girder/cumulus/server/models/cluster.py]: Should this be done in Girder filter method?

File: cumulus/constants.py [hunk 7]
```diff
@@ -18,14 +18,20 @@
 ###############################################################################
 
 
+from enum import IntEnum
+
+
 class ClusterType:
     EC2 = 'ec2'
+    ANSIBLE = 'ansible'
     TRADITIONAL = 'trad'
```
Comment (cjh1) [inline cumulus/constants.py]: I do have concerns about introducing a cluster type of 'ansible'.  In the future when we support 'openstack' and 'ec2' how will be distinguish between the two? May be we need the concept of a cluster "provisioner"

File: cumulus/constants.py [hunk 8]
```diff
@@ -54,3 +60,20 @@ class JobQueueState:
...
+    created = 10
+    launching = 20
+    launched = 30
+    provisioning = 40
+    provisioned = 50
+    terminating = 60
+    terminated = 70
+    stopped = 101
+    running = 102
+
```
Comment (cjh1) [inline cumulus/constants.py]: If we are going to move toward using enum for the states we should do the same for job etc. in the future.

**Comment by cjh1** [comment]
Can we document that variable that we can pass in when creating an "ansible" cluster? For starcluster we currently had:

cluster_size
master_instance_type
node_instance_type
node_image_id
master_image_id

We also need to rebase on currently master

File: cumulus/constants.py (hunk 7 above)
Comment (kotfic) [inline cumulus/constants.py]: I agree,  having a type of 'ansible' is just a stop gap until we can get feature parity with starcluster then (hopefully) pull out starcluster and replace the naming convention

File: girder/cumulus/server/cluster.py (hunk 2 above)
Comment (cjh1) [inline girder/cumulus/server/cluster.py]: So should we rename it 'variables" or "playbookVariables"?

**Comment by cjh1** [comment]
Looking good, can we update ansible_integration_test.py and make use it still works?

**Comment by cjh1** [comment]
@kotfic I am going to go ahead a merge this, I have raise #165 for the testing issue I am seeing.
